### PR TITLE
feat(firestore-palm-chatbot): add palm provider option to chatbot

### PR DESCRIPTION
**Pull request by cabljac**
this PR:
- adds an option to select Vertex AI as provider of the PaLM API for the chatbot extension
- adds the option to use an API key with GL PaLM API
- adds the codechat model as an option for the vertex provider

### CODE REVIEW

File: firestore-palm-chatbot/functions/src/discussion.ts [hunk 1]
```diff
@@ -15,107 +15,29 @@
...
 
 export class Discussion {
-  private client: DiscussServiceClient;
+  private generativeClient?: DiscussServiceClient;
+  private vertexClient?: v1.PredictionServiceClient;
+  private endpoint?: string;
   context?: string;
   examples?: Message[] = [];
-  model = 'models/chat-bison-001';
+  model = config.useVertex ? 'chat-bison@001' : 'models/chat-bison-001';
```
Comment (dackers86) [inline firestore-palm-chatbot/functions/src/discussion.ts:40]: The above config has removed the `models/` is that also needed here?

File: firestore-palm-chatbot/functions/src/discussion.ts [hunk 2]
```diff
@@ -202,6 +206,44 @@ export class Discussion {
...
+  ): Promise<GenerateMessageResponse> {
+    if (!this.vertexClient) {
+      throw new Error('Vertex client not initialized.');
+    }
+
+    const [result] = await this.vertexClient.predict(request);
+
+    const prediction = result.predictions![0];
+
+    // TODO: fix type casting
```
Comment (dackers86) [inline firestore-palm-chatbot/functions/src/discussion.ts:220]: Are the TODOs still needed?

File: firestore-palm-chatbot/functions/src/discussion.ts (hunk 1 above)
Comment (cabljac) [reply firestore-palm-chatbot/functions/src/discussion.ts:40]: yeah for vertex it doesn't have the `models` prefix

File: firestore-palm-chatbot/functions/src/discussion.ts (hunk 2 above)
Comment (cabljac) [inline firestore-palm-chatbot/functions/src/discussion.ts:220]:
```suggestion
```

Comment (cabljac) [reply firestore-palm-chatbot/functions/src/discussion.ts:220]: can remove this, i think the casting is fine for now as the Google Types are not very illuminative

**Review by dackers86 — APPROVED** [review]

**Review by dackers86 — APPROVED** [review]

**Pull request merged**